src/System.Security.Permissions/src/System/Security/Permissions/PrincipalPermission.cs

```diff
@@ -2,25 +2,471 @@
+            }
+
+            PrincipalPermission operand = (PrincipalPermission)target;
+
+            if (operand.IsUnrestricted())
+            {
+                return Copy();
+            }
+
+            List<IDRole> idroles = null;
+
+            for (int i = 0; i < m_array.Length; ++i)
+            {
+                for (int j = 0; j < operand.m_array.Length; ++j)
+                {
+                    if (operand.m_array[j].m_authenticated == m_array[i].m_authenticated)
+                    {
+                        if (operand.m_array[j].m_id == null ||
+                            m_array[i].m_id == null ||
+                            m_array[i].m_id.Equals(operand.m_array[j].m_id))
+                        {
+                            if (idroles == null)
+                            {
+                                idroles = new List<IDRole>();
+                            }
+
+                            IDRole idrole = new IDRole();
+
+                            idrole.m_id = operand.m_array[j].m_id == null ? m_array[i].m_id : operand.m_array[j].m_id;
+
+                            if (operand.m_array[j].m_role == null ||
+                                m_array[i].m_role == null ||
+                                m_array[i].m_role.Equals(operand.m_array[j].m_role))
+                            {
+                                idrole.m_role = operand.m_array[j].m_role == null ? m_array[i].m_role : operand.m_array[j].m_role;
+                            }
+                            else
+                            {
+                                idrole.m_role = "";
+                            }
+
+                            idrole.m_authenticated = operand.m_array[j].m_authenticated;
+
+                            idroles.Add(idrole);
+                        }
+                        else if (operand.m_array[j].m_role == null ||
+                                 m_array[i].m_role == null ||
+                                 m_array[i].m_role.Equals(operand.m_array[j].m_role))
+                        {
+                            if (idroles == null)
+                            {
+                                idroles = new List<IDRole>();
+                            }
+
+                            IDRole idrole = new IDRole();
+
+                            idrole.m_id = "";
+                            idrole.m_role = operand.m_array[j].m_role == null ? m_array[i].m_role : operand.m_array[j].m_role;
+                            idrole.m_authenticated = operand.m_array[j].m_authenticated;
+
+                            idroles.Add(idrole);
+                        }
+                    }
+                }
+            }
+
+            if (idroles == null)
+            {
+                return null;
+            }
+            else
+            {
+                IDRole[] idrolesArray = new IDRole[idroles.Count];
+
+                IEnumerator idrolesEnumerator = idroles.GetEnumerator();
```

Comment:
This is going to allocate a boxed enumerator, instead of using List's struct enumerator. But this entire implementation could simply be:

```c#
return new PrincipalPermission(idroles.ToArray());
```